rfd/0096-helm-chart-revamp.md

```diff
@@ -0,0 +1,338 @@
+- `authentication`
+- `authenticationType`
+- `authenticationSecondFactor`
+- `proxyListenerMode`
+- `sessionRecording`
+- `separatePostgresListener`
+- `separateMongoListener`
+- `kubePublicAddr`
+- `mongoPublicAddr`
+- `mysqlPublicAddr`
+- `postgresPublicAddr`
+- `sshPublicAddr`
+- `tunnelPublicAddr`
+- `acme`
+- `acmeEmail`
+- `acmeURI`
+- `log`
+
+#### Chart-specific configuration values
+
+Some values are used to configure the Kubernetes resources deploying Teleport.
+When specified they should apply to both auth and proxy deployments.  Those
+values are:
+
+- `podSecurityPolicy`
+- `labels`
+- `highAvailability`
+- `tls`
+- `image`
+- `enterpriseImage`
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `imagePullPolicy`
+- `initContainers`
+- `postStart`
+- `securityContext`
+- `priorityClassName`
+- `tolerations`
+- `probeTimeoutSeconds`
+- `teleportVersionOverride`
+- `resources`
+
+A few values will have to be treated differently:
+
+- `persistence` will only apply to the `auth` deployment
+- `service` will only apply to the `proxy` service
+- `serviceAccount.name` will apply to the `auth`, the proxy service account
+  name should be the auth one suffixed with `-proxy`
+
+Some users will need to set different values for auth and proxy pods, the
+following values should be also available under `auth` and `proxy`. Those
+specific values should take precedence over the ones at the root.
+
+- `labels`
+- `highAvailability` (except the certManager section)
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `initContainers`
+- `postStart`
+- `tolerations`
+- `teleportVersionOverride`
+- `resources`
+
+### Update strategy between major versions
+
+Auth pods have to be updated before proxies.  Helm does not support applying
+resources in a specific order.
+
+Both auth and proxy rollouts will be triggered at the same time, but the proxy
+one should be held until all auth pods are rolled out. Not waiting for the full
+rollout will cause the load to spread unevenly across auth pods, which will be
+harmful at scale.
+
+Proxies will have an initContainer checking if all auth pods from the past
+version were removed. Version check via the Teleport gRPC api (`PingResponse`)
+requires valid credentials to connect to Teleport. To workaround this issue we
+can rely on Kubernetes's service discovery through DNS to discover how many
+pods are running which version:
+
+- the chart labels auth pods with their major teleport version
+- the chart creates two headless services:
+  - one selecting pods with the current major version `teleport-auth-v11`
+  - one selecting pods with the previous major version `teleport-auth-v10`
+- proxy pods have an initContainer
+- the `v11` initContainer resolves `teleport-auth-v10` every 5 seconds until no
+  IP is returned
+- the initContainer exits, the proxy starts
+- this unlocks the proxy deployment rollout
```

Comment:
This is unsafe with a StatefulSet because a pod killed during a rollout will be re-created with the lastest version. Deployments implementation is based on ReplicaSets, which ensures this does not happen and we cannot have a capacity loss during a stuck rollout.

I wonder if it would be possible for the chart to refuse `sessionRecording: proxy`. In the last chart, running in `proxy` mode was almost identical to running in `auth` mode (almost because if users manually added proxies it would have had a different behaviour). I believe that in order to improve the overall experience we have to provide a more opinionated deployment of Teleport, one that works great in 99% of the use-cases.

Reply:
I discussed with @r0mant and it seems not supporting `proxy` session recording mode does not align with our strategy. However we could gracefully handle session recording queue emptying. This would mean the `proxy` session recording would only loose data on hardware failure, not on controlled disruptions like autoscaling events or version updates/configuration rollouts.

Reply:
After discussing with multiple persons, it appears graceful shutdown is not currently supported by the chart, and does not even ship the recordings before exiting. Working around with a graceful restart also raises issues because thet restart will hang until users session end, but proxy is not preemptively closing user sessions.

Reply:
Might be good to investigate if we can use the  `lifecycle.preStop` hook.
The exec plugin could force teleport to close connections and ship the data volume to other proxy nodes to avoid the loss of session recording inventory.

Using a statefulset it's possible to know if your replica will be scaled down -> using the number of replicas and the pod name. For cases where scale down will happen, we can ship the payload to another proxy

Reply:
> Might be good to investigate if we can use the `lifecycle.preStop` hook. The exec plugin could force teleport to close connections and ship the data volume to other proxy nodes to avoid the loss of session recording inventory.
> 
> Using a statefulset it's possible to know if your replica will be scaled down -> using the number of replicas and the pod name. For cases where scale down will happen, we can ship the payload to another proxy

As far as I understand the proxy storage is temporary, even in `proxy` mode. The recordings will eventually be shipped to an auth server for archiving purposes. In a permanent downscale case we would need to have a way to forcefully empty the queue. I initially assumed people using `proxy` mode would not downscale, but this was not based and this case is valid.

My biggest concern is what happens if we're in a stuck proxy rollout (auths are slowly rolling out, proxies are waiting) and we loose one or many proxies. With a `StatefulSet` they will come back up as `v11` and not `v10`. They will not go back up and wait for the auth rollout to finish. This will cause a capacity loss, and maybe a full blackout if all proxies gets rescheduled (e.g. a GKE automated nodepool update that happens at the same time).

Reply:
I changed back to a `Deployment`, this means users might loose active session recordings during a pod rollout if using the `proxy` recording mode.

Reply:
IIRC the way we do this in Cloud is to rewrite the default container's `SIGTERM` (15) signal to `SIGQUIT` (3) using `dumb-init` so that we do in fact ask the pods to shut down gracefully: `/usr/bin/dumb-init --rewrite 15:3 -- teleport start -c /etc/teleport/teleport.yaml`

(from https://goteleport.com/docs/reference/signals/):
`QUIT`: Graceful shutdown. The daemon will wait until connections are dropped.
`TERM` , `INT` or `KILL`: Immediate non-graceful shutdown. All existing connections will be dropped.

I've never proposed this change in the Teleport Docker container itself because I feel it would be too much of a breaking change, but doing manually inside the Helm chart with args/command might work and be a good compromise?

I know that someone on the Teleport side was investigating the ability for Teleport to properly flush its session recording queue as part of a graceful shutdown but I'm unsure what progress they made on that. Cloud doesn't support proxy recording mode so there isn't a ton of prior art unfortunately.

Reply:
Yeah, @espadolini  took a lot of time to explain me this (thank you ❤️ ). I don't think I want to/can dive this rabbit hole right now: `SIGQUIT` waits for sessions to end, which can take several hours as we don't have a session eviction/migration mechanism. In cloud this kinda works because we can afford slow rollouts with 30min of grace period. I'd love to raise this issue again once we have done the main refactos.